utils/notification_service_doc_tests.py

```diff
@@ -133,21 +133,38 @@ def failures(self) -> Dict:
 
     @property
     def category_failures(self) -> Dict:
+        MAX_ERROR_TEXT = 3000 - len("The following examples had failures:\n\n\n\n")
         line_length = 40
         category_failures = {k: v["failed"] for k, v in doc_test_results.items() if isinstance(v, dict)}
 
-        report = ""
-        for category, failures in category_failures.items():
+        def single_category_failures(category, failures, target_max_len, report):
+            text = ""
             if len(failures) == 0:
-                continue
+                return ""
 
             if report != "":
-                report += "\n\n"
+                text += "\n"
+
+            text += f"*{category} failures*:".ljust(line_length // 2).rjust(line_length // 2) + "\n"
+
+            for idx, failure in enumerate(failures):
+                new_text = text
+                new_text += f"`{failure}`\n"
+                if len(new_text) > target_max_len - len("[Truncated]\n"):
+                    text = text + "[Truncated]\n"
+                    break
+                text = new_text
 
-            report += f"*{category} failures*:".ljust(line_length // 2).rjust(line_length // 2) + "\n"
-            report += "`"
-            report += "`\n`".join(failures)
-            report += "`"
+            return text
+
+        report = ""
+        for category, failures in category_failures.items():
+            max_len = MAX_ERROR_TEXT - len(report) - len("\n[Truncated]")
+            new_text = single_category_failures(category, failures, target_max_len=max_len, report=report)
+            if len(new_text) > max_len:
+                report += "\n[Truncated]"
+                break
+            report += new_text
```

Comment:
A bit more complex than before: here we have 2 for loops: outer loop for categories, and inner loop for each single failure in a specific category.